builder/dockerfile/shell_parser.go

```diff
@@ -73,36 +84,23 @@ func (sw *shellWord) processStopOn(stopChar rune) (string, error) {
 	return result, nil
 }
 
-func (sw *shellWord) peek() rune {
-	if sw.pos == len(sw.word) {
-		return '\000'
-	}
-	return rune(sw.word[sw.pos])
-}
-
-func (sw *shellWord) next() rune {
-	if sw.pos == len(sw.word) {
-		return '\000'
-	}
-	ch := rune(sw.word[sw.pos])
-	sw.pos++
-	return ch
-}
-
 func (sw *shellWord) processSingleQuote() (string, error) {
 	// All chars between single quotes are taken as-is
 	// Note, you can't escape '
 	var result string
 
-	sw.next()
+	if sw.scanner.Next() == scanner.EOF {
```

Comment:
I think this will never be true - I think Next() will always return the starting single quote - which is why we just ignored the return in the old code. So I think you can remove the "if" check and just call sw.scanner.Next().